### PR TITLE
Unify lwAFTR and other state query around YANG

### PR DESCRIPTION
**Pull request by wingo**
This PR renames the `state.show_state` function to `state.read_state`, and makes it return a value that follows the corresponding state grammar; notably, this state value (or subvalue inside it) is a cdata object if all state values are of fixed-sized types and have defaults.

It also converts `snabb lwaftr query` and `snabb lwaftr check` to use a helper from `lwcounter` to turn YANG state values into counter tables, reading the set of available counters from the YANG schema instead of an embedded list.  This will free things up so that we can get rid of the central `lwcounter` state and instead use the SHM frame mechanism for instantiating counters for apps.

### CODE REVIEW

File: src/program/lwaftr/query/query.lua [hunk 1]
```diff
@@ -24,8 +23,24 @@ local function sort (t)
    return t
 end
 
-local function is_counter_name (name)
-   return lwcounter.counter_names[name] ~= nil
+local function counter_names ()
```
Comment (dpino) [inline src/program/lwaftr/query/query.lua]: This function is duplicated in check/util.lua. Likely is worth to put it somewhere like in `apps/lwaftr/lwutil.lua`.

File: src/program/lwaftr/check/util.lua [hunk 2]
```diff
@@ -18,13 +15,24 @@ function load_requested_counters(counters)
    return result
 end
 
+local function counter_names()
+   local names = {}
+   local schema = schema.load_schema_by_name('snabb-softwire-v2')
+   for k, node in pairs(schema.body['softwire-state'].body) do
```
Comment (dpino) [inline src/program/lwaftr/check/util.lua]: What do you think about taking advantage of this PR to put counters into their own container? I would need to adapt this code when alarms get added. I can live without it for now though.

File: src/apps/config/leader.lua [hunk 3]
```diff
@@ -558,17 +558,18 @@ end
...
    local function getter()
       if args.schema ~= self.schema_name then
-            return self:foreign_rpc_get_state(args.schema, args.path, args)
+         return self:foreign_rpc_get_state(args.schema, args.path,
+                                           args.format, args.print_default)
       end
       local printer = path_printer_for_schema_by_name(
          self.schema_name, args.path, false, args.format, args.print_default)
-      local s = {}
+      local states = {}
```
Comment (teknico) [inline src/apps/config/leader.lua:566]: Thank you. :-)

File: src/apps/lwaftr/lwcounter.lua [hunk 4]
```diff
@@ -13,123 +17,31 @@ local shm = require("core.shm")
...
-   "in-ipv6-frag-needs-reassembly",
-   "in-ipv6-frag-reassembled",
-   "in-ipv6-frag-reassembly-unneeded",
-   "drop-ipv6-frag-disabled",
-   "drop-ipv6-frag-invalid-reassembly",
-   "drop-ipv6-frag-random-evicted",
-   "out-ipv6-frag",
-   "out-ipv6-frag-not",
-   "memuse-ipv6-frag-reassembly-buffer",
+function counter_names ()
```
Comment (teknico) [inline src/apps/lwaftr/lwcounter.lua:21]: All those painstakingly crafted names, sigh. Where have they gone? ;-)

File: src/lib/yang/state.lua [hunk 5]
```diff
@@ -5,23 +5,23 @@ local lib = require("core.lib")
...
 
-local function flatten(val)
-   local rtn = {}
+local function flatten(val, out)
+   out = out or {}
    for k, v in pairs(val) do
       if type(v) == "table" then
-         v = flatten(v)
-         for k1, v1 in pairs(v) do rtn[k1] = v1 end
+         flatten(v, out)
```
Comment (teknico) [inline src/lib/yang/state.lua:16]: Nice optimization.

File: src/program/lwaftr/query/query.lua [hunk 6]
```diff
@@ -80,14 +67,17 @@ local function print_counter (name, value, max_width)
    print(("%s: %s%s"):format(name, (" "):rep(nspaces), lib.comma_value(value)))
 end
 
-local function print_counters (tree, filter)
+local function print_counters (pid, filter)
    print("lwAFTR operational counters (non-zero)")
    -- Open, read and print whatever counters are in that directory.
-   local counters, max_width = read_counters(tree)
+   local counters = lwcounter.read_counters(pid)
+   local max_width = max_key_width(counters)
```
Comment (teknico) [inline src/program/lwaftr/query/query.lua:74]: Nice separation of concerns.

**Review by teknico — APPROVED** [review]
Nice changes. The compilery stuff in state.lua flew a little over my head, though.

**Pull request merged**